### PR TITLE
Replace Ice_invokeResult by a tuple

### LINKED ISSUE #1828: Replace Ice::Object::Ice_invokeResult by tuple
This is the last remaining "result struct". 

We should replace it by `std::tuple<bool, std::vector<std::byte>>`.

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR replaces Ice_invokeResult by a tuple and performs other related clean-ups.

Fixes #1828.

### CODE REVIEW

File: cpp/include/Ice/Object.h [hunk 1]
```diff
@@ -101,7 +101,6 @@ class ICE_API Object
      */
     static const std::string& ice_staticId();
 
-// TODO: temporary. To remove it, we need to refactor IncomingAsync that is the base class for generated AMD callbacks.
```
Comment (bernardnormier) [inline cpp/include/Ice/Object.h:104]: I believe this TODO was an accidental leftover from the C++98 mapping removal.

File: cpp/include/Ice/OutgoingAsync.h [hunk 2]
```diff
@@ -562,113 +562,6 @@ class PromiseOutgoing<P, void> : public OutgoingAsyncT<void>, public PromiseInvo
     }
 };
 
-template<typename R>
```
Comment (bernardnormier) [inline cpp/include/Ice/OutgoingAsync.h:565]: Moved these classes as-is to ProxyAsync.cpp

File: cpp/include/Ice/Proxy.h [hunk 3]
```diff
@@ -336,18 +335,6 @@ class ICE_API ObjectPrx : public Proxy<ObjectPrx>
     ObjectPrx& operator=(const ObjectPrx& rhs) noexcept = default;
     ObjectPrx& operator=(ObjectPrx&& rhs) noexcept = default;
 
-    /**
```
Comment (bernardnormier) [inline cpp/include/Ice/Proxy.h:339]: Moved towards the bottom of the class

File: cpp/include/Ice/Proxy.h [hunk 4]
```diff
@@ -485,15 +472,6 @@ class ICE_API ObjectPrx : public Proxy<ObjectPrx>
     _iceI_id(const std::shared_ptr<::IceInternal::OutgoingAsyncT<std::string>>&, const Ice::Context&) const;
     /// \endcond
 
-    /**
-     * Returns the Slice type ID associated with this type.
-     * @return The Slice type ID.
-     */
-    static const std::string& ice_staticId()
```
Comment (bernardnormier) [inline cpp/include/Ice/Proxy.h:492]: Moved towards the bottom of the class.

File: cpp/include/Ice/Proxy.h [hunk 5]
```diff
@@ -601,6 +579,66 @@ class ICE_API ObjectPrx : public Proxy<ObjectPrx>
                     const Ice::Context& context = Ice::noExplicitContext) const;
 
     /**
+     * Obtains the Connection for this proxy. If the proxy does not yet have an established connection,
```
Comment (bernardnormier) [inline cpp/include/Ice/Proxy.h:582]: Moved up to keep all invocations / async functions together.

File: cpp/include/Ice/Proxy.h [hunk 6]
```diff
@@ -763,64 +801,22 @@ class ICE_API ObjectPrx : public Proxy<ObjectPrx>
     bool ice_isFixed() const;
 
     /**
-     * Obtains the Connection for this proxy. If the proxy does not yet have an established connection,
```
Comment (bernardnormier) [inline cpp/include/Ice/Proxy.h:766]: Moved down after the invocation / async functions.

File: cpp/src/Ice/ProxyAsync.cpp [hunk 7]
```diff
@@ -139,6 +139,113 @@ class ProxyFlushBatchPromise : public ProxyFlushBatchAsync, public PromiseInvoke
...
+                         std::function<void(bool)> sent) :
+        InvokeOutgoingAsyncT<R>(proxy, false), LambdaInvoke(std::move(ex), std::move(sent))
+    {
+        if(response)
+        {
+            _response = [this, response = std::move(response)](bool ok)
+            {
+                if(this->_is.b.empty())
+                {
+                    response(R { ok, { 0, 0 }});
```
Comment (pepone) [inline cpp/src/Ice/ProxyAsync.cpp:204]: Should we apply the 3.7 fix, or you want to further refactor this code?

https://github.com/zeroc-ice/ice/commit/49f99d1dcc042a800840e1814b6a3a5f24c1ebb4

File: cpp/src/Ice/ProxyAsync.cpp [hunk 8]
```diff
@@ -139,6 +139,113 @@ class ProxyFlushBatchPromise : public ProxyFlushBatchAsync, public PromiseInvoke
...
+    InvokePromiseOutgoing(const Ice::ObjectPrx& proxy, bool synchronous) :
+        InvokeOutgoingAsyncT<R>(proxy, false)
+    {
+        this->_synchronous = synchronous;
+        this->_response = [this](bool ok)
+        {
+            if(this->_is.b.empty())
+            {
+                std::vector<Ice::Byte> encaps;
+                this->_promise.set_value(R { ok, encaps});
```
Comment (pepone) [inline cpp/src/Ice/ProxyAsync.cpp:229]:
```suggestion
                this->_promise.set_value(R { ok, std::move(encaps) });
```
?

File: cpp/src/Ice/ProxyAsync.cpp [hunk 9]
```diff
@@ -139,6 +139,113 @@ class ProxyFlushBatchPromise : public ProxyFlushBatchAsync, public PromiseInvoke
...
+            }
+        };
+    }
+
+    virtual bool handleSent(bool done, bool) override
+    {
+        if(done)
+        {
+            std::vector<Ice::Byte> encaps;
+            this->_promise.set_value(R { true, encaps});
```
Comment (pepone) [inline cpp/src/Ice/ProxyAsync.cpp:243]: ditto, can we move this?
```suggestion
            this->_promise.set_value(R { true, std::move(encaps) });
```

**Review by pepone — APPROVED** [review]
Looks good!

File: cpp/src/Ice/ProxyAsync.cpp (hunk 7 above)
Comment (bernardnormier) [reply cpp/src/Ice/ProxyAsync.cpp:204]: I plan to refactor this code much more.

File: cpp/src/Ice/ProxyAsync.cpp (hunk 8 above)
Comment (bernardnormier) [reply cpp/src/Ice/ProxyAsync.cpp:229]: Pretty minor since it's an empty vector.

File: cpp/src/Ice/ProxyAsync.cpp (hunk 9 above)
Comment (bernardnormier) [reply cpp/src/Ice/ProxyAsync.cpp:243]: Same here

**Review by externl — APPROVED** [review]
Looks good!

**Pull request merged**